### ISSUE TITLE
Consider reusing StringBuilders across event writes

### ISSUE THREAD

**Issue by epignosisx** [NLog/NLog#1140]
Have you consider reusing StringBuilders across event writes? 

NLog makes sure to minimize string allocations by sharing a StringBuilder for all involved layout renderers. However, could take the savings a level up by introducing a StringBuilderCache like this one:

https://github.com/opserver/Opserver/blob/8a18ff8b8550ea8d5793580477ecb0f7aaa3dfee/Opserver.Core/StringBuilderCache.cs

It caches the StringBuilder per thread and it has a shared StringBuilder for cases where a nested method requests the cached StringBuilder but it was taking by a method higher up in the stacktrace.

**Comment by 304NotModified** [comment]
Hi!

This sounds like a cool idea! I like it, but we need some well timing tests for this. It should be only merged when the improvement is significant.

**Comment by 304NotModified** [comment]
Alex,

Thanks for all the proposals and code contributions! We are really happy with your style of code and contributions. We think it would be great if you would like to join the NLog team! Check https://github.com/NLog/NLog/issues/828 for some info. Just let us know!

**Comment by epignosisx** [comment]
Thanks for the invitation! Yes, I would like to join the team.

**Comment by epignosisx** [comment]
Totally agree about carefully testing the reuse of StringBuilderCache. I'll start doing some tests.

**Comment by 304NotModified** [comment]
:) Cool! Will send you an invite when I'm behind a pc.

**Comment by epignosisx** [comment]
I did some testing using the StringBuilderCache and just using it in SimpleLayout like this:

```c#
protected override string GetFormattedMessage(LogEventInfo logEvent)
{
    if (IsFixedText)
    {
        return this.fixedText;
    }

    string cachedValue;

    if (logEvent.TryGetCachedLayoutValue(this, out cachedValue))
    {
        return cachedValue;
    }

    int initialSize = this.maxRenderedLength;
    if (initialSize > MaxInitialRenderBufferLength)
    {
        initialSize = MaxInitialRenderBufferLength;
    }

    //var builder = new StringBuilder(initialSize);
    var builder = StringBuilderCache.Get(initialSize); //<-- change here

    //Memory profiling pointed out that using a foreach-loop was allocating
    //an Enumerator. Switching to a for-loop avoids the memory allocation.
    for (int i = 0; i < this.Renderers.Count; i++)
    {
        LayoutRenderer renderer = this.Renderers[i];
        try
        {
            renderer.Render(builder, logEvent);
        }
        catch (Exception exception)
        {
            if (exception.MustBeRethrown())
            {
                throw;
            }

            if (InternalLogger.IsWarnEnabled)
            {
                InternalLogger.Warn("Exception in {0}.Append(): {1}.", renderer.GetType().FullName, exception);
            }
        }
    }

    if (builder.Length > this.maxRenderedLength)
    {
        this.maxRenderedLength = builder.Length;
    }

    //string value = builder.ToString();
    string value = builder.ToStringRecycle(); //<-- change here
    logEvent.AddCachedLayoutValue(this, value);
    return value;
}
```

and used this app to check memory usage:

```c#
static void Main(string[] args)
{
    LogManager.Configuration = CreateConfigurationFromString(@"
        <nlog>
            <targets>
                <target name='Debug' type='Debug' layout='${longdate}|${level}|${message}'>
                </target>
            </targets>
            <rules>
                <logger name='*' levels='Debug' writeTo='Debug'></logger>
            </rules>
        </nlog>
    ");

    ILogger logger = LogManager.GetLogger("A");
    logger.Log(LogLevel.Debug, "message");

    Console.WriteLine("Take First snapshot");
    Console.ReadLine();
    
    for (int i = 0; i < 100; i++)
    {
        logger.Log(LogLevel.Debug, "message");
    }

    Console.WriteLine("Take Last snapshot");
    Console.ReadLine();
}
```

As expected memory consumption went down from:

![image](https://cloud.githubusercontent.com/assets/697377/12224541/8a47df5c-b7c1-11e5-9826-87c7c7108c04.png)

to:
![image](https://cloud.githubusercontent.com/assets/697377/12224542/959dd5a0-b7c1-11e5-8f7e-b19c8ed07bc5.png)

Notice how **char[]** dropped from first offender. Also the total number of bytes and objects went down.

Then, I did a performance test with the following app:

```c#
static void Main(string[] args)
{
    LogManager.Configuration = CreateConfigurationFromString(@"
        <nlog>
            <targets>
                <target name='Debug' type='Debug' layout='${longdate}|${level}|${message}'>
                </target>
            </targets>
            <rules>
                <logger name='*' levels='Debug' writeTo='Debug'></logger>
            </rules>
        </nlog>
    ");

    ILogger logger = LogManager.GetLogger("A");
    logger.Log(LogLevel.Debug, "message");

    var sw = new Stopwatch();
    for (int j = 0; j < 25; j++)
    {
        sw.Restart();
        for (int i = 0; i < 1000000; i++)
        {
            logger.Log(LogLevel.Debug, "message");
        }
        sw.Stop();
        Console.WriteLine(sw.ElapsedMilliseconds);
    }
}
```

and the numbers favor the original code:

|        | Original (ms) | Cache (ms)  |
|--------|----------|--------|
| Run 1  | 1093     | 1150   |
| Run 2  | 1094     | 1150   |
| Run 3  | 1089     | 1150   |
| Run 4  | 1098     | 1151   |
| Run 5  | 1104     | 1152   |
| Run 6  | 1129     | 1153   |
| Run 7  | 1101     | 1154   |
| Run 8  | 1097     | 1152   |
| Run 9  | 1095     | 1147   |
| Run 10 | 1093     | 1150   |
| Run 11 | 1094     | 1148   |
| Run 12 | 1090     | 1165   |
| Run 13 | 1099     | 1149   |
| Run 14 | 1095     | 1152   |
| Run 15 | 1089     | 1145   |
| Run 16 | 1091     | 1144   |
| Run 17 | 1092     | 1158   |
| Run 18 | 1096     | 1178   |
| Run 19 | 1097     | 1151   |
| Run 20 | 1091     | 1167   |
| Run 21 | 1090     | 1148   |
| Run 22 | 1116     | 1144   |
| Run 23 | 1089     | 1150   |
| Run 24 | 1089     | 1149   |
| Run 25 | 1091     | 1148   |
| **Avg**    | **1096.08**  | **1152.2** |

An increase of 0.00005612 ms per log write. I consider this increase to be small enough. The memory savings would help reduce memory pressure. So for long running apps trying maintain high throughput this change could be beneficial.

**Comment by snakefoot** [comment]
Similar strategy is applied with #1663

**Comment by snakefoot** [comment]
Maybe resolved with #1799 ? (NLog 4.4.2)

**Comment by 304NotModified** [comment]
Yes indeed! Thanks!
